File: src/content/docs/topics/cs/leetcode-150/trees/1448-count-good-nodes-in-binary-tree.py
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def build_tree(vals):
    if not vals:
        return None
    root = TreeNode(vals[0])
    q = [root]
    i = 1
    while q and i < len(vals):
        node = q.pop(0)
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            q.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            q.append(node.right)
        i += 1
    return root

def good_nodes(root):
    pass  # TODO: implement

def _run_tests():
    assert good_nodes(build_tree([3, 1, 4, 3, None, 1, 5])) == 4
    assert good_nodes(build_tree([3, 3, None, 4, 2])) == 3
    assert good_nodes(build_tree([1])) == 1
    assert good_nodes(build_tree([5, 4, 6, 3, None, None, 7])) == 3
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/trees/1448-count-good-nodes-in-binary-tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func goodNodes(root *TreeNode) int {
	// TODO: implement
	return 0
}

func runTests() {
	// [3,1,4,3,null,1,5] -> 4
	root1 := &TreeNode{Val: 3,
		Left:  &TreeNode{Val: 1, Left: &TreeNode{Val: 3}},
		Right: &TreeNode{Val: 4, Left: &TreeNode{Val: 1}, Right: &TreeNode{Val: 5}},
	}
	assert(goodNodes(root1) == 4)

	// [3,3,null,4,2] -> 3
	root2 := &TreeNode{Val: 3,
		Left: &TreeNode{Val: 3, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 2}},
	}
	assert(goodNodes(root2) == 3)

	// single node -> 1
	assert(goodNodes(&TreeNode{Val: 1}) == 1)

	// [5,4,6,3,null,null,7] -> 3 (5, 6, 7 are good)
	root4 := &TreeNode{Val: 5,
		Left:  &TreeNode{Val: 4, Left: &TreeNode{Val: 3}},
		Right: &TreeNode{Val: 6, Right: &TreeNode{Val: 7}},
	}
	assert(goodNodes(root4) == 3)

	fmt.Println("all tests pass")
}

func main() { runTests() }
